development/code/error-handling.md

```diff
@@ -0,0 +1,112 @@
+# Error handling guidelines
+
+## Primary rule
+
+**Undefined behavior is absolutely never acceptable.**
+
+- Methods such as `panic` / `abort` / `unwrap` / `expect` are **not** undefined behavior.
+- Such methods are still best to avoid where reasonable error handling is possible.
+
+## Guidelines
+
+- **Use error types**
+  - **Prefer returning a [`Result`] type rather than panicking**.
+    - This rule is especially true in library code.
+    - This rule is less strictly required for application code but still typically recommended.
+
+  - More generally, **use dedicated error types**.
+
+    `enums` or `structs` decorated with the `#[derive(Debug, thiserror::Error )]` attribute are extremely helpful for
+    error handling.
+    They allow the compiler to enforce exhaustive matching on error types, and massively assist in downstream
+    refactoring when the number or type of possible errors changes.
+
+    Any deviation from this pattern should be
+    - carefully considered,
+    - explicitly justified in documentation,
+    - and treated with skepticism and prejudice on later review.
+
+  - **Option** is not a good error
+
+    Avoid using `Option` as an error type.
+    Instead, use a dedicated error type that indicates the absence of a value is an error.
+    Even using `Result<T, ()>` is preferable to `Option` as it suggests an error rather than a missing, but optional
+    value.
+
+  - **String errors**
+
+    Arbitrary `String` or "string like" errors are **strongly discouraged** and must only be used where absolutely
+    necessary (such as when they are imposed on us by an external framework).
+    Such errors are _actively hostile_ to correct and sustainable error handling.
```

Comment:
You don't explain why and this makes it sound like a subjective opinion. I'd probably avoid such considerations in general.

Overall, I'm unsure you need to explain the reasoning behind each style choice. I'd assume an AI tool to understand why typed errors are preferable, without going too much into the details.